core/services/address/CountrySubRegionDao.php

```diff
@@ -0,0 +1,122 @@
+<?php
+
+namespace EventEspresso\core\services\address;
+
+use EE_Country;
+use EE_Error;
+use EEM_State;
+use EventEspresso\core\exceptions\InvalidDataTypeException;
+use EventEspresso\core\exceptions\InvalidInterfaceException;
+use InvalidArgumentException;
+use ReflectionException;
+use stdClass;
+
+/**
+ * Class CountrySubRegionDao
+ * Data Access Object for retrieving Country and Country SubRegion data
+ *
+ * @package EventEspresso\core\services\address
+ * @author  Brent Christensen
+ * @since   $VID:$
+ */
+class CountrySubRegionDao
+{
+
+    const REPO_URL = 'https://raw.githubusercontent.com/eventespresso/countries-and-provinces-states-regions/master/';
+
+    /**
+     * @var array $countries
+     */
+    private $countries = array();
+
+
+    /**
+     * @param EE_Country $country_object
+     * @return void
+     * @throws EE_Error
+     * @throws InvalidArgumentException
+     * @throws InvalidDataTypeException
+     * @throws InvalidInterfaceException
+     * @throws ReflectionException
+     */
+    public function saveCountrySubRegions(EE_Country $country_object)
+    {
+        $CNT_ISO = $country_object->ID();
+        $has_sub_regions = EEM_State::instance()->count(array(array('Country.CNT_ISO' => $CNT_ISO)));
+        if ($has_sub_regions) {
+            return;
+        }
+        if (empty($this->countries)) {
+            $this->countries = $this->retrieveJsonData(self::REPO_URL . 'countries.json');
+        }
+        if (is_array($this->countries)) {
+            foreach ($this->countries as $key => $country) {
+                if ($country instanceof stdClass
+                    && $country->code === $CNT_ISO
+                    && empty($country->sub_regions)
+                    && $country->filename !== null
+                ) {
+                    $country->sub_regions = $this->retrieveJsonData(
+                        self::REPO_URL . 'countries/' . $country->filename . '.json'
+                    );
+                    $this->saveSubRegionData($country, $country->sub_regions);
+                }
+            }
+        }
+    }
+
+
+    /**
+     * @param string $url
+     * @return array
+     */
+    private function retrieveJsonData($url)
+    {
+        if (empty($url)) {
+            return array();
+        }
+        $request = wp_safe_remote_get($url);
+        if (is_wp_error($request)) {
+            return array();
+        }
+        $body = wp_remote_retrieve_body($request);
+        return json_decode($body);
+    }
```

Comment:
How often will this data change?  Since we've forked it, I wonder if we should have the json file include a version property and then the initial request can check for version and compare with the version from any previous fetch?  So when a fetch is performed we save the json file version somewhere locally.  That way in future iterations if we want to expose a ui/ux for event managers to be notified that country/region data is updated and provide the opportunity for them to update locally we can do that?

Reply:
hmmm... ya... was trying to just get a solution out the door... but if we don't account for that, then there isn't really any way to update this without going in and nuking the db... sigh...

Reply:
kk added versioning to the JSON data: https://github.com/eventespresso/countries-and-subregions/blob/master/countries.json